hera_sim/visibilities/vis_cpu.py

```diff
@@ -0,0 +1,328 @@
+    vis_cpu visibility simulator.
+
+    This is a fast, simple visibility simulator that is intended to be
+    replaced by vis_gpu. It extends :class:`VisibilitySimulator`.
+    """
+
+    def __init__(self, bm_pix=100, real_dtype=np.float32,
+                 complex_dtype=np.complex64, **kwargs):
+        """
+        Parameters
+        ----------
+        bm_pix : int, optional
+            The number of pixels along a side in the beam map when
+            converted to (l, m) coordinates. Defaults to 100.
+        real_dtype : {np.float32, np.float64}
+            Data type for real-valued arrays.
+        complex_dtype : {np.complex64, np.complex128}
+            Data type for complex-valued arrays.
+        **kwargs
+            Arguments of :class:`VisibilitySimulator`.
+        """
+
+        self._real_dtype = real_dtype
+        self._complex_dtype = complex_dtype
+        self.bm_pix = bm_pix
+
+        super(VisCPU, self).__init__(**kwargs)
+
+        # Convert some arguments to forms more simple for vis_cpu.
+        self.antpos = self.uvdata.get_ENU_antpos()[0].astype(self._real_dtype)
+        self.freqs = self.uvdata.freq_array[0]
+
+    @property
+    def lsts(self):
+        """
+        Sets LSTs from uvdata if not already set.
+
+        Returns
+        -------
+        array_like
+            LSTs of observations. Shape=(NTIMES,).
+        """
+        try:
+            return self.__lsts
+        except AttributeError:
+            self.__lsts = self.uvdata.lst_array[::self.uvdata.Nbls]
+
+            return self.__lsts
+
+    def validate(self):
+        """Checks for correct input format."""
+        super(VisCPU, self).validate()
+
+        # This one in particular requires that every baseline is used!
+        N = len(self.uvdata.antenna_numbers)
+        # N(N-1)/2 unique cross-correlations + N autocorrelations.
+        if len(self.uvdata.get_antpairs()) != N * (N + 1) / 2:
+            raise ValueError("VisCPU requires using every pair of antennas, "
+                             "but the UVData object does not comply.")
+
+        if (len(self.uvdata.data_array) != len(self.uvdata.get_antpairs())
+                * len(self.lsts)):
+            raise ValueError("VisCPU requires that every baseline uses the "
+                             "same LSTS.")
+
+    def get_beam_lm(self):
+        """
+        Obtain the beam pattern in (l,m) co-ordinates for each antenna.
+
+        Returns
+        -------
+        array_like
+            The beam pattern in (l,m) for each antenna.
+            Shape=(NANT, BM_PIX, BM_PIX).
+
+        Notes
+        -----
+            Due to using the verbatim :func:`vis_cpu` function, the beam
+            cube must have an entry for each antenna, which is a bit of
+            a waste of memory in some cases. If this is changed in the
+            future, this method can be modified to only return one
+            matrix for each beam.
+        """
+        return np.array([
```

Comment:
If you're concerned about memory, then you might want to use `np.asarray` instead. `np.array` defaults to make a copy of the object to be made into an `ndarray` (you could also just do `copy=False`, but I think using `np.asarray` is cleaner).

Reply:
Cool, thanks, didn't know that!